benchmark/_driver.py

```diff
@@ -0,0 +1,139 @@
+# Copyright 2015 ClusterHQ Inc.  See LICENSE file for details.
+"""
+Driver for the control service benchmarks.
+"""
+
+from eliot import start_action
+from eliot.twisted import DeferredContext
+
+from twisted.python.filepath import FilePath
+from twisted.internet.task import cooperate
+from twisted.internet.defer import maybeDeferred, Deferred, succeed
+
+from flocker.apiclient import FlockerClient
+
+
+def benchmark(metric, operation, scenario, num_samples=3):
+    """
+    Perform sampling of the operation.
+
+    :param IMetric metric: A quantity to measure.
+    :param IOperation operation: An operation to perform.
+    :param IScenario scenario: A load scenario.
+    :param int num_samples: Number of samples to take.
+    :return: Deferred firing with a list of samples. Each sample is a
+        dictionary containing a ``success`` boolean. If ``success is True``,
+        the dictionary also contains a ``value`` for the sample measurement.
+        If ``success is False``, the dictionary also contains a ``reason`` for
+        failure.
+    """
+    scenario_established = scenario.start()
+
+    samples = []
+
+    def sample(i):
+        with start_action(action_type=u'flocker:benchmark:sample', sample=i):
+            sampling = DeferredContext(maybeDeferred(operation.get_probe))
+
+            def run_probe(probe):
+                probing = metric.measure(probe.run)
+                probing.addCallbacks(
+                    lambda interval: samples.append(
+                        dict(success=True, value=interval)
+                    ),
+                    lambda reason: samples.append(
+                        dict(success=False, reason=reason.getTraceback()),
+                    ),
+                )
+                probing.addCallback(lambda ignored: probe.cleanup())
+                return probing
+            sampling.addCallback(run_probe)
+            sampling.addActionFinish()
+            return sampling.result
```

Comment:
`benchmark` is a fairly large function. I think it'd be clearer if `sample` were extracted to a top-level, private function that is called by `collect_samples`. Maybe (i.e. optionally) it would benefit from direct tests?